File: tools/mlflow_fun/tools/dump_experiment.py

"""
Recursively dumps all information about an experiment including all details of its runs and their params, metrics and artifacts.
Note that this can be expensive. Adjust your artifact_max_level.
"""

import mlflow
from mlflow_fun.tools import dump_run

client = mlflow.tracking.MlflowClient()
print("MLflow Version:", mlflow.version.VERSION)
print("MLflow Tracking URI:", mlflow.get_tracking_uri())

def dump_experiment(exp_id_or_name, artifact_max_level, show_info, show_data):
    print("Options:")
    print("  exp_id_or_name:",exp_id_or_name)
    print("  artifact_max_level:",artifact_max_level)
    print("  show_info:",show_info)
    print("  show_data:",show_data)

    if exp_id_or_name.isdigit():
        exp = client.get_experiment(exp_id_or_name)
        which = "ID"
    else:
        exp = client.get_experiment_by_name(exp_id_or_name)
        which = "name"
    if exp is None:
         raise Exception("Cannot find experiment {} '{}'".format(which,exp_id_or_name))
    exp_id = exp.experiment_id
    print("experiment_id:",exp_id)
    dump_experiment_details(exp)

    if show_info or show_data:
        infos = client.list_run_infos(exp_id)
        print("  #runs:",len(infos))
        #dump_runs(infos, artifact_max_level)
        print("Runs:")
        for j,info in enumerate(infos):
            print("  Run {}/{}:".format(j+1,len(infos)))
            if show_data:
                dump_run.dump_run_id(info.run_uuid, artifact_max_level, indent="    ")
            else:
                dump_run.dump_run_info(info, indent="    ")

def dump_experiment_details(exp):
    print("Experiment Details:")
    for k,v in exp.__dict__.items(): print("  {}: {}".format(k[1:],v))
  
def dump_runs(infos, artifact_max_level):
    print("Runs:")
    for j,info in enumerate(infos):
        print("  Run {}/{}:".format(j+1,len(infos)))
        dump_run.dump_run_id(info.run_uuid, artifact_max_level, indent="    ")

if __name__ == "__main__":
    from argparse import ArgumentParser
    parser = ArgumentParser()
    parser.add_argument("--experiment_id_or_name", dest="experiment_id", help="Experiment ID", required=True)
    parser.add_argument("--artifact_max_level", dest="artifact_max_level", help="Number of artifact levels to recurse", required=False, default=1, type=int)
    parser.add_argument("--show_info", dest="show_info", help="Show run info", required=False, default=False, action='store_true')
    parser.add_argument("--show_data", dest="show_data", help="Show run info and data", required=False, default=False, action='store_true')
    args = parser.parse_args()
    print("args:",args)
    dump_experiment(args.experiment_id, args.artifact_max_level,args.show_info, args.show_data)

File: examples/spark-scala/src/main/scala/org/andre/mlflow/tools/DumpExperiment.scala
package org.andre.mlflow.tools

import scala.collection.JavaConversions._
import com.beust.jcommander.{JCommander, Parameter}
import org.mlflow.tracking.MlflowClient
import org.andre.mlflow.util.MLflowUtils

object DumpExperiment {
  def dumpExperiment(client: MlflowClient, experimentId: String, artifactMaxLevel: Int, showRunInfo: Boolean, showRunData: Boolean, inDatabricks: Boolean) {
    val expResponse = client.getExperiment(experimentId)
    val exp1 = expResponse.getExperiment
    println(s"Experiment Details:")
    println(s"  experimentId: ${exp1.getExperimentId}")
    println(s"  name: ${exp1.getName}")
    println(s"  artifactLocation: ${exp1.getArtifactLocation}")
    println(s"  lifecycleStage: ${exp1.getLifecycleStage}")
    println(s"  runsCount: ${expResponse.getRunsCount}")

    println(s"Runs:")
    if (showRunInfo || showRunData) {
        val infos = if (inDatabricks) client.listRunInfos(experimentId) else expResponse.getRunsList
        for((info,j) <- infos.zipWithIndex) {
          println(s"  Run ${j+1}/${infos.size}:")
          if (showRunData) {
            DumpRun.dumpRun(client, info.getRunId, artifactMaxLevel,"    ")
          } else {
            DumpRun.dumpRunInfo(info,"    ")
          }
        }
    }
  }

  def main(args: Array[String]) {
    new JCommander(opts, args.toArray: _*)
    println("Options:")
    println(s"  tracking URI: ${opts.trackingUri}")
    println(s"  token: ${opts.token}")
    println(s"  experimentId: ${opts.experimentId}")
    println(s"  artifactMaxLevel: ${opts.artifactMaxLevel}")
    println(s"  showRunInfo: ${opts.showRunInfo}")
    println(s"  showRunData: ${opts.showRunData}")
    println(s"  inDatabricks: ${opts.inDatabricks}")
    val client = MLflowUtils.createMlflowClient(opts.trackingUri, opts.token)
    dumpExperiment(client, opts.experimentId, opts.artifactMaxLevel, opts.showRunInfo, opts.showRunData, opts.inDatabricks)
  }

  object opts {
    @Parameter(names = Array("--trackingUri" ), description = "Tracking Server URI", required=false)
    var trackingUri: String = null

    @Parameter(names = Array("--token" ), description = "REST API token", required=false)
    var token: String = null

    @Parameter(names = Array("--experimentId" ), description = "rexperimentId", required=true)
    var experimentId: String = null

    @Parameter(names = Array("--artifactMaxLevel" ), description = "Number of artifact levels to recurse", required=false)
    var artifactMaxLevel = 1

    @Parameter(names = Array("--showRunInfo" ), description = "Show run info", required=false)
    var showRunInfo = false

    @Parameter(names = Array("--showRunData" ), description = "Show run info and data", required=false)
    var showRunData = false

    @Parameter(names = Array("--inDatabricks" ), description = "Running inside Databricks", required=false)
    var inDatabricks = false
  }
}
